src/encoded/audit/experiment.py

```diff
@@ -3195,50 +3195,58 @@ def get_file_accessions(list_of_files):
     return accessions_set
 
 
-def is_outdated_bams_replicate(bam_file, files_structure):
-
+def is_outdated_bams_replicate(bam_file, files_structure, assay_name):
+    # if derived_from contains accessions that were not in
+    # original_files and not in contributing files - it is outdated!    
     for file_id in bam_file.get('derived_from'):
         if file_id not in files_structure.get('original_files') and \
            file_id not in files_structure.get('contributing_files'):
             return True
 
-    # if derived_from contains accessions that were not in
-    # original_files and not in contributing files - it is outdated!
     derived_from_fastqs = get_derived_from_files_set(
         [bam_file], files_structure, 'fastq', True)
 
+
     # if there are no FASTQs we can not find our the replicate
     if len(derived_from_fastqs) == 0:
         return False
 
     derived_from_fastq_accessions = get_file_accessions(derived_from_fastqs)
 
-    bio_rep = []
-    for fastq_file in derived_from_fastqs:
-        if 'biological_replicates' in fastq_file and \
-           len(fastq_file['biological_replicates']) != 0:
-            for entry in fastq_file['biological_replicates']:
-                bio_rep.append(entry)
-            break
+    rep_fastqs = []
+    replicate_type = 'biological_replicates'
+    rep = bam_file.get('biological_replicates')
+    # for DNase we should consider technial relicates
+    if assay_name != 'ChIP-seq':
+        replicate_type = 'technical_replicates'
+        rep = bam_file.get('technical_replicates')
+    
+    # number of replicates BAM file should belong to have to be one
+    # in cases where it is more than one, there probably was replicates 
+    # reorganization, that invalidates the analysis    
+    if rep and len(rep) > 1:
+        return True
 
-    bio_rep_fastqs = []
     for fastq_file in files_structure.get('fastq_files').values():
-        if 'biological_replicates' in fastq_file:
-            for entry in fastq_file['biological_replicates']:
-                if entry in bio_rep:
-                    bio_rep_fastqs.append(fastq_file)
+        if replicate_type in fastq_file:
```

Comment:
I believe this can be rewritten in a cleaner/faster way.    First it filters the fastq_files by the replicate type.  Then, using sets and a generator, sees if there is a match between bam_file replicate types and fastq_file replicate types.
```
rep_type_fastqs = [
    f for f in files_structure.get('fastq_files').values()
    if replicate_type in f
]
rep_set = set(rep)
rep_fastqs2 = [
    f for f in rep_type_fastqs
    if any(e in rep_set for e in set(f[replicate_type]))
]
```
When lists are cast to sets duplicates are removed.  I don't think that effects the outcome?

Reply:
done